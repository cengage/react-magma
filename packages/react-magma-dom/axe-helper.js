import { configureAxe } from 'jest-axe';

export const axe = configureAxe({
  rules: {
    // disabled landmark rules when testing isolated components.
    region: { enabled: false },
    // disabled nested-interactive when testing isolated components
    'nested-interactive': { enabled: false },
  },
});
