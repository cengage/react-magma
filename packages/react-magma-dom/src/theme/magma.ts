import { ThemeTransitions, transitions } from './components/transition';

export interface Colors {
  primary: string;
  primary100: string;
  primary200: string;
  primary300: string;
  primary400: string;
  primary500: string;
  primary600: string;
  primary700: string;

  secondary: string;
  secondary500: string;
  secondary600: string;
  secondary700: string;

  tertiary: string;
  tertiary500: string;
  tertiary600: string;
  tertiary700: string;

  neutral: string;
  neutral100: string;
  neutral200: string;
  neutral300: string;
  neutral400: string;
  neutral500: string;
  neutral600: string;
  neutral700: string;
  neutral800: string;
  neutral900: string;

  info: string;
  info100: string;
  info200: string;
  info300: string;
  info400: string;
  info500: string;
  info600: string;
  info700: string;

  danger: string;
  danger100: string;
  danger200: string;
  danger300: string;
  danger400: string;
  danger500: string;
  danger600: string;
  danger700: string;

  warning: string;
  warning100: string;
  warning200: string;
  warning300: string;
  warning400: string;
  warning500: string;
  warning600: string;
  warning700: string;

  success: string;
  success100: string;
  success200: string;
  success300: string;
  success400: string;
  success500: string;
  success600: string;
  success700: string;

  focus: string;
  focusInverse: string;

  border: string;
  borderInverse: string;
}

export interface Breakpoints {
  xs: number;
  small: number;
  medium: number;
  large: number;
  xl: number;
}

export interface IconSizes {
  xSmall: number;
  small: number;
  medium: number;
  large: number;
  xLarge: number;
}

export interface SpacingScale {
  spacing01: string;
  spacing02: string;
  spacing03: string;
  spacing04: string;
  spacing05: string;
  spacing06: string;
  spacing07: string;
  spacing08: string;
  spacing09: string;
  spacing10: string;
  spacing11: string;
  spacing12: string;
  spacing13: string;
  spacing14: string;
}

export interface TypographyExpressiveSizes {
  heading2XLarge: any;
  headingXLarge: any;
  headingLarge: any;
  headingMedium: any;
  headingSmall: any;
  headingXSmall: any;
  heading2XSmall: any;
  bodyLarge: any;
  bodyMedium: any;
  bodySmall: any;
  bodyXSmall: any;
}

export interface TypeScaleSize {
  fontSize: string;
  letterSpacing?: string;
  lineHeight: string;
}

export interface TypeScale {
  size01: TypeScaleSize;
  size02: TypeScaleSize;
  size03: TypeScaleSize;
  size04: TypeScaleSize;
  size05: TypeScaleSize;
  size06: TypeScaleSize;
  size07: TypeScaleSize;
  size08: TypeScaleSize;
  size09: TypeScaleSize;
  size10: TypeScaleSize;
  size11: TypeScaleSize;
  size12: TypeScaleSize;
  size13: TypeScaleSize;
  size14: TypeScaleSize;
  size15: TypeScaleSize;
}

export interface VisualStyle {
  mobile: TypeScaleSize;
  desktop: TypeScaleSize;
  fontWeight?: number;
}

export interface TypographyVisualStyles {
  headingXLarge: VisualStyle;
  headingLarge: VisualStyle;
  headingMedium: VisualStyle;
  headingSmall: VisualStyle;
  headingXSmall: VisualStyle;
  heading2XSmall: VisualStyle;
  bodyLarge: VisualStyle;
  bodyMedium: VisualStyle;
  bodySmall: VisualStyle;
  bodyXSmall: VisualStyle;
}

export interface TypographyVisualStylesExpressive {
  heading2XLarge: VisualStyle;
  headingXLarge: VisualStyle;
  headingLarge: VisualStyle;
  headingMedium: VisualStyle;
  headingSmall: VisualStyle;
  headingXSmall: VisualStyle;
  heading2XSmall: VisualStyle;
  bodyLarge: VisualStyle;
  bodyMedium: VisualStyle;
  bodySmall: VisualStyle;
  bodyXSmall: VisualStyle;
}

interface AppBar {
  backgroundColor: string;
  height: string;
  padding: string;
  textColor: string;
  compact: {
    height: string;
    padding: string;
  };
  inverse: {
    backgroundColor: string;
    textColor: string;
  };
}
export interface Modal {
  width: {
    small: string;
    medium: string;
    large: string;
  };
}

interface Drawer {
  default: {
    maxWidth: string;
    minHeight: string;
    borderRadius: string;
    margin: string;
    top: string;
    bottom: string;
    left: string;
    right: string;
    position: string;
  };
  right: {
    left: string;
    height: string;
    width: string;
  };
  left: {
    right: string;
    height: string;
    width: string;
  };
  top: {
    bottom: string;
    height: string;
  };
  bottom: {
    top: string;
    height: string;
  };
}

interface Combobox {
  menu: {
    maxHeight: string;
  };
}

interface Dropdown {
  content: {
    maxHeight: string;
  };
}

interface Dropdown {
  content: {
    maxHeight: string;
  };
}

interface Select {
  menu: {
    maxHeight: string;
  };
}

interface Tabs {
  approxTabSize: {
    horizontal: number;
    vertical: number;
  };
}

export interface Tooltip {
  arrowSize: string;
  arrowSizeDoubled: string;
  backgroundColor: string;
  fontWeight: number | string;
  maxWidth: string;
  textColor: string;
  typeScale: TypeScaleSize;
  zIndex: number;
  inverse: any;
}

export interface ThemeInterface {
  appBar: AppBar;
  breakpoints: Breakpoints;
  bodyFont: string;
  bodyExpressiveFont: string;
  bodyNarrativeFont: string;
  borderRadius: string;
  borderRadiusSmall: string;
  colors: Colors;
  combobox: Combobox;
  direction: string;
  drawer: Drawer;
  dropdown: Dropdown;
  headingFont: string;
  iconSizes: IconSizes;
  iterableColors: string[];
  chartColors?: string[];
  chartColorsInverse?: string[];
  spacingMultiplier: number;
  spaceScale: SpacingScale;
  headingExpressiveFont: string;
  headingNarrativeFont: string;
  select: Select;
  typeScale: TypeScale;
  typographyVisualStyles: TypographyVisualStyles;
  typographyExpressiveVisualStyles: TypographyVisualStylesExpressive;
  typographyNarrativeVisualStyles: TypographyVisualStyles;
  modal: Modal;
  tabs: Tabs;
  tooltip: Tooltip;
  transitions: ThemeTransitions;
}

const typeScale = {
  size01: {
    fontSize: '12px',
    letterSpacing: '.32px',
    lineHeight: '16px',
  },
  size02: {
    fontSize: '14px',
    letterSpacing: '.16px',
    lineHeight: '20px',
  },
  size03: {
    fontSize: '16px',
    letterSpacing: '.32px',
    lineHeight: '24px',
  },
  size04: {
    fontSize: '18px',
    lineHeight: '24px',
  },
  size05: {
    fontSize: '20px',
    lineHeight: '32px',
  },
  size06: {
    fontSize: '24px',
    lineHeight: '32px',
  },
  size07: {
    fontSize: '28px',
    lineHeight: '40px',
  },
  size08: {
    fontSize: '32px',
    lineHeight: '40px',
  },
  size09: {
    fontSize: '36px',
    lineHeight: '48px',
  },
  size10: {
    fontSize: '42px',
    lineHeight: '48px',
  },
  size11: {
    fontSize: '48px',
    lineHeight: '56px',
  },
  size12: {
    fontSize: '52px',
    lineHeight: '68px',
  },
  size13: {
    fontSize: '54px',
    lineHeight: '64px',
  },
  size14: {
    fontSize: '60px',
    lineHeight: '72px',
  },
  size15: {
    fontSize: '72px',
    lineHeight: '84px',
  },
};

const primaryColors = {
  primary100: '#E8E9F8',
  primary200: '#BABDE9',
  primary300: '#8B91DA',
  primary400: '#5D65CB',
  primary500: '#3942B0',
  primary600: '#292F7C',
  primary700: '#1A1E51',
};

const secondaryColors = {
  secondary500: '#FEE449',
  secondary600: '#FEDA0D',
  secondary700: '#ECC901',
};

const tertiaryColors = {
  tertiary500: '#CDDEFF',
  tertiary600: '#A2C1FF',
  tertiary700: '#79A5FF',
};

const neutralColors = {
  neutral100: '#FFFFFF', // white
  neutral200: '#F5F5F5',
  neutral300: '#D4D4D4',
  neutral400: '#8D8D8D',
  neutral500: '#707070',
  neutral600: '#5A5A5A',
  neutral700: '#454545',
  neutral800: '#2D2D2D',
  neutral900: '#000000', // black
};

const infoColors = {
  info100: '#E8F5FC',
  info200: '#A6DEFF',
  info300: '#2FB3FF',
  info400: '#009AF3',
  info500: '#0074B7',
  info600: '#005F96',
  info700: '#004A75',
};

const dangerColors = {
  danger100: '#FDEFEE',
  danger200: '#FAAEB0',
  danger300: '#E8716D',
  danger400: '#E24943',
  danger500: '#D32821',
  danger600: '#A91F1A',
  danger700: '#7F1714',
};

const warningColors = {
  warning100: '#FCEEE5',
  warning200: '#F6CDB2',
  warning300: '#E98B4C',
  warning400: '#E06A1C',
  warning500: '#AD5115',
  warning600: '#8D4311',
  warning700: '#6E340E',
};

const successColors = {
  success100: '#E3FAEA',
  success200: '#ACF0C1',
  success300: '#3EDD6E',
  success400: '#21B94E',
  success500: '#178037',
  success600: '#136A2D',
  success700: '#0F5323',
};

const colors = {
  primary: primaryColors.primary500,
  secondary: secondaryColors.secondary500,
  tertiary: tertiaryColors.tertiary500,
  neutral: neutralColors.neutral500,
  info: infoColors.info500,
  danger: dangerColors.danger500,
  warning: warningColors.warning500,
  success: successColors.success500,
  ...primaryColors,
  ...secondaryColors,
  ...tertiaryColors,
  ...neutralColors,
  ...infoColors,
  ...dangerColors,
  ...warningColors,
  ...successColors,

  focus: infoColors.info500,
  focusInverse: infoColors.info200,

  border: neutralColors.neutral300,
  borderInverse: 'rgba(255,255,255,0.25)',
};

const spaceScale = {
  spacing01: '2px',
  spacing02: '4px',
  spacing03: '8px',
  spacing04: '12px',
  spacing05: '16px',
  spacing06: '24px',
  spacing07: '28px',
  spacing08: '32px',
  spacing09: '40px',
  spacing10: '48px',
  spacing11: '56px',
  spacing12: '64px',
  spacing13: '96px',
  spacing14: '160px',
};

export const magma = {
  bodyFont: '"Work Sans",Helvetica,sans-serif',
  bodyExpressiveFont: '"Work Sans",Helvetica,sans-serif',
  bodyNarrativeFont: "'Noto Serif',Times New Roman,serif",
  borderRadius: '8px',
  borderRadiusSmall: '4px',
  colors: colors,
  headingFont: '"Work Sans",Helvetica,sans-serif',
  headingExpressiveFont: '"Work Sans",Helvetica,sans-serif',
  headingNarrativeFont: "'Noto Serif',Times New Roman,serif",
  direction: 'ltr',
  spacingMultiplier: 8,
  transitions,
  // breakpoints
  breakpoints: {
    xs: 0,
    small: 600,
    medium: 768,
    large: 1024,
    xl: 1200,
  },

  spaceScale: spaceScale,

  iconSizes: {
    xSmall: 16,
    small: 20,
    medium: 24,
    large: 32,
    xLarge: 54,
  },

  // Typography
  typeScale: typeScale,
  typographyVisualStyles: {
    // Productive
    headingXLarge: {
      mobile: typeScale.size07,
      desktop: typeScale.size09,
      fontWeight: 600,
    },
    headingLarge: {
      mobile: typeScale.size06,
      desktop: typeScale.size07,
      fontWeight: 600,
    },
    headingMedium: {
      mobile: typeScale.size05,
      desktop: typeScale.size06,
      fontWeight: 600,
    },
    headingSmall: {
      mobile: typeScale.size04,
      desktop: typeScale.size05,
      fontWeight: 600,
    },
    headingXSmall: {
      mobile: typeScale.size04,
      desktop: typeScale.size04,
      fontWeight: 600,
    },
    heading2XSmall: {
      mobile: typeScale.size01,
      desktop: typeScale.size01,
      fontWeight: 700,
    },
    bodyLarge: {
      mobile: typeScale.size04,
      desktop: typeScale.size05,
    },
    bodyMedium: {
      mobile: typeScale.size03,
      desktop: typeScale.size03,
    },
    bodySmall: {
      mobile: typeScale.size02,
      desktop: typeScale.size02,
    },
    bodyXSmall: {
      mobile: typeScale.size01,
      desktop: typeScale.size01,
    },
  },
  typographyExpressiveVisualStyles: {
    heading2XLarge: {
      mobile: typeScale.size13,
      desktop: typeScale.size15,
      fontWeight: 600,
    },
    headingXLarge: {
      mobile: typeScale.size10,
      desktop: typeScale.size12,
      fontWeight: 400,
    },
    headingLarge: {
      mobile: typeScale.size07,
      desktop: typeScale.size09,
      fontWeight: 300,
    },
    headingMedium: {
      mobile: typeScale.size06,
      desktop: typeScale.size07,
      fontWeight: 300,
    },
    headingSmall: {
      mobile: typeScale.size05,
      desktop: typeScale.size06,
      fontWeight: 300,
    },
    headingXSmall: {
      mobile: typeScale.size04,
      desktop: typeScale.size05,
      fontWeight: 300,
    },
    heading2XSmall: {
      mobile: typeScale.size03,
      desktop: typeScale.size03,
      fontWeight: 700,
    },
    bodyLarge: {
      mobile: typeScale.size05,
      desktop: typeScale.size06,
    },
    bodyMedium: {
      mobile: typeScale.size03,
      desktop: typeScale.size03,
    },
    bodySmall: {
      mobile: typeScale.size02,
      desktop: typeScale.size02,
    },
    bodyXSmall: {
      mobile: typeScale.size01,
      desktop: typeScale.size01,
    },
  },
  typographyNarrativeVisualStyles: {
    headingXLarge: {
      mobile: typeScale.size07,
      desktop: typeScale.size08,
      fontWeight: 700,
    },
    headingLarge: {
      mobile: typeScale.size06,
      desktop: typeScale.size07,
      fontWeight: 700,
    },
    headingMedium: {
      mobile: typeScale.size05,
      desktop: typeScale.size06,
      fontWeight: 700,
    },
    headingSmall: {
      mobile: typeScale.size04,
      desktop: typeScale.size05,
      fontWeight: 700,
    },
    headingXSmall: {
      mobile: typeScale.size04,
      desktop: typeScale.size04,
      fontWeight: 700,
    },
    heading2XSmall: {
      mobile: typeScale.size01,
      desktop: typeScale.size01,
      fontWeight: 700,
    },
    bodyLarge: {
      mobile: typeScale.size04,
      desktop: typeScale.size05,
    },
    bodyMedium: {
      mobile: typeScale.size03,
      desktop: typeScale.size03,
    },
    bodySmall: {
      mobile: typeScale.size02,
      desktop: typeScale.size02,
    },
    bodyXSmall: {
      mobile: typeScale.size01,
      desktop: typeScale.size01,
    },
  },

  appBar: {
    backgroundColor: colors.neutral100,
    height: '88px',
    padding: `${spaceScale.spacing06} ${spaceScale.spacing05}`,
    textColor: colors.neutral,
    compact: {
      height: '56px',
      padding: `${spaceScale.spacing05} ${spaceScale.spacing05} ${spaceScale.spacing05} ${spaceScale.spacing06}`,
    },
    inverse: {
      backgroundColor: colors.primary600,
      textColor: colors.neutral100,
    },
  },

  combobox: {
    menu: {
      maxHeight: '250px',
    },
  },

  modal: {
    width: {
      small: '300px',
      medium: '600px',
      large: '900px',
    },
  },

  drawer: {
    default: {
      maxWidth: '100%',
      minHeight: '300px',
      margin: '0',
      borderRadius: '0',
      top: '0',
      right: '0',
      bottom: '0',
      left: '0',
      position: 'relative',
    },
    right: {
      left: 'auto',
      height: '100%',
      width: '300px',
      position: 'fixed',
    },
    left: {
      right: 'auto',
      height: '100%',
      width: '300px',
    },
    top: {
      bottom: 'auto',
      height: '300px',
    },
    bottom: {
      top: 'auto',
      height: '300px',
      position: 'fixed',
    },
  },

  dropdown: {
    content: {
      maxHeight: '250px',
    },
  },

  iterableColors: [
    '#0085CC',
    '#E0004D',
    '#FA6600',
    '#48A200',
    '#B12FAD',
    '#00A393',
    '#00507A',
    '#8F0033',
    '#B84900',
    '#255200',
    '#711E6E',
    '#005249',
  ],

  chartColors: [
    '#009AF3',
    '#E0004D',
    '#1EA746',
    '#FA6600',
    '#B12FAD',
    '#00A393',
    '#005F96',
    '#8F0033',
    '#136A2D',
    '#B84900',
    '#711E6E',
    '#005249',
  ],

  chartColorsInverse: [
    '#1FB0FF',
    '#FF337A',
    '#65E000',
    '#FF9147',
    '#D45ED0',
    '#00E0CA',
    '#85D4FF',
    '#FF99BD',
    '#FFB685',
    '#C7FF99',
    '#E9AFE7',
    '#99FFF5',
  ],

  tabs: {
    approxTabSize: {
      horizontal: 120,
      vertical: 80,
    },
  },

  select: {
    menu: {
      maxHeight: '250px',
    },
  },

  tag: {
    border: '0',
    display: 'inline-flex',
    alignItems: 'center',
    justifyContent: 'space-around',
  },

  tooltip: {
    arrowSize: '4px',
    arrowSizeDoubled: '8px',
    backgroundColor: colors.neutral700,
    fontWeight: 500,
    maxWidth: '300px',
    textColor: colors.neutral100,
    typeScale: typeScale.size01,
    zIndex: 999,
    inverse: {
      backgroundColor: colors.neutral100,
      textColor: colors.neutral700,
    },
  },
};
