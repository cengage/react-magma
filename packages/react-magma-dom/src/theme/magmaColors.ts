export enum MagmaColors {
  primary = 'primary',
  secondary = 'secondary',
  tertiary = 'tertiary',
  neutral = 'neutral',
  info = 'info',
  danger = 'danger',
  warning = 'warning',
  success = 'success',
}
