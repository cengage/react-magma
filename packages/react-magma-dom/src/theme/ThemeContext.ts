import * as React from 'react';

import { magma } from './magma';

export const ThemeContext = React.createContext(
  magma // default value
);
