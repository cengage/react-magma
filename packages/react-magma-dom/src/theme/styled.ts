import styled from '@emotion/styled-base';
import { CreateStyled } from '@emotion/styled/types';
import { ThemeInterface } from './magma';

export default styled as CreateStyled<ThemeInterface>;
