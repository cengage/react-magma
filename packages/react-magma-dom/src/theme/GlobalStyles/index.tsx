import * as React from 'react';
import { Global, css } from '@emotion/core';
import { ThemeContext } from '../ThemeContext';
import { useIsInverse } from '../../inverse';

function getStyles(theme, isInverse: boolean) {
  return css`
    @import url('https://fonts.googleapis.com/css2?family=Work+Sans:ital,wght@0,100;0,200;0,300;0,400;0,500;0,600;0,700;0,800;0,900;1,100;1,200;1,300;1,400;1,500;1,600;1,700;1,800;1,900&display=swap');
    @import url('https://fonts.googleapis.com/css2?family=Noto+Serif:ital,wght@0,400;0,700;1,400;1,700&display=swap');

    *,
    *:before,
    *:after {
      box-sizing: border-box;
    }

    *:focus {
      outline: 2px solid
        ${isInverse ? theme.colors.focusInverse : theme.colors.focus};
      outline-offset: 2px;
    }

    html {
      -ms-text-size-adjust: 100%;
      -webkit-text-size-adjust: 100%;
      font-size: ${theme.typeScale.size03.fontSize};
      line-height: ${theme.typeScale.size03.lineHeight};
      scroll-behavior: smooth;
    }

    html,
    body {
      background: ${isInverse
        ? theme.colors.foundation
        : theme.colors.neutral08};
      color: ${isInverse ? theme.colors.neutral08 : theme.colors.neutral};
      margin: 0;
      padding: 0;
    }

    body {
      font-family: ${theme.bodyFont};
      font-style: normal;
      font-weight: 400;
      font-size: ${theme.typeScale.size03.fontSize};
      line-height: ${theme.typeScale.size03.lineHeight};
    }

    a {
      color: ${isInverse ? theme.colors.neutral07 : theme.colors.primary};
      cursor: pointer;
      text-decoration: underline;

      &:hover,
      &:focus {
        color: ${isInverse ? theme.colors.neutral06 : theme.colors.focus};
      }
    }

    button,
    input,
    optgroup,
    select,
    textarea {
      font: inherit;
      margin: 0;
    }

    svg:not(:root) {
      overflow: hidden;
    }
  `;
}

export const GlobalStyles: React.FunctionComponent = () => {
  const isInverse = useIsInverse();
  return (
    <ThemeContext.Consumer>
      {theme => <Global styles={getStyles(theme, isInverse)} />}
    </ThemeContext.Consumer>
  );
};
