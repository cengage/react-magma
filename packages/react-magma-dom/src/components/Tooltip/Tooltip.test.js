/* eslint-disable no-console */
import React from 'react';
import { axe } from '../../../axe-helper';
import { Tooltip } from '.';
import { act, render, fireEvent } from '@testing-library/react';
import { magma } from '../../theme/magma';

const CONTENT_TEXT = 'Test Content';
const TRIGGER_ELEMENT = <button>Test trigger</button>;

describe('Tooltip', () => {
  it('should find element by testId', () => {
    const testId = 'test-id';
    const { getByTestId } = render(
      <Tooltip content="content" testId={testId}>
        <button />
      </Tooltip>
    );

    expect(getByTestId(testId)).toBeInTheDocument();
  });

  it('should render the tooltip component, positioned top by default', async () => {
    const { container, getByText } = render(
      <Tooltip id="tooltipID" content={CONTENT_TEXT}>
        {TRIGGER_ELEMENT}
      </Tooltip>
    );
    const tooltipTrigger = container.querySelector('button');

    expect(container).toBeInTheDocument();

    expect(tooltipTrigger).toBeInTheDocument();

    await act(async () => {
      tooltipTrigger.focus();
    });

    const tooltip = container.querySelector('div[role="tooltip"]');
    const arrow = container.querySelector('span');

    expect(tooltip).toBeInTheDocument();
    expect(arrow).toBeInTheDocument();

    expect(tooltip).toHaveAttribute('data-popper-placement', 'top');

    expect(tooltip).toMatchSnapshot();
  });

  it('should render the tooltip component with the correct styles when positioned left', async () => {
    const { container } = render(
      <Tooltip position="left" content={CONTENT_TEXT}>
        {TRIGGER_ELEMENT}
      </Tooltip>
    );
    const tooltipTrigger = container.querySelector('button');

    await act(async () => {
      tooltipTrigger.focus();
    });

    const tooltip = container.querySelector('div[role="tooltip"]');

    expect(tooltip).toHaveAttribute('data-popper-placement', 'left');
  });

  it('should render the tooltip component with the correct styles when positioned right', async () => {
    const { container } = render(
      <Tooltip position="right" content={CONTENT_TEXT}>
        {TRIGGER_ELEMENT}
      </Tooltip>
    );
    const tooltipTrigger = container.querySelector('button');

    await act(async () => {
      tooltipTrigger.focus();
    });

    const tooltip = container.querySelector('div[role="tooltip"]');

    expect(tooltip).toHaveAttribute('data-popper-placement', 'right');
  });

  it('should render the tooltip component with the correct styles when positioned bottom', async () => {
    const { container } = render(
      <Tooltip position="bottom" content={CONTENT_TEXT}>
        {TRIGGER_ELEMENT}
      </Tooltip>
    );
    const tooltipTrigger = container.querySelector('button');

    await act(async () => {
      tooltipTrigger.focus();
    });

    const tooltip = container.querySelector('div[role="tooltip"]');

    expect(tooltip).toHaveAttribute('data-popper-placement', 'bottom');
  });

  it('should show the tooltip on focus and hide it on blur', () => {
    const { container, getByText } = render(
      <Tooltip content={CONTENT_TEXT}>{TRIGGER_ELEMENT}</Tooltip>
    );
    const tooltipTrigger = container.querySelector('button');

    expect(
      container.querySelector('div[role="tooltip"]')
    ).not.toBeInTheDocument();
    fireEvent.focus(tooltipTrigger);

    expect(container.querySelector('div[role="tooltip"]')).toBeInTheDocument();
    fireEvent.blur(tooltipTrigger);
  });

  it('should allow for a persistent tooltip that does not hide on blur', () => {
    const { container, getByText } = render(
      <Tooltip open content={CONTENT_TEXT}>
        {TRIGGER_ELEMENT}
      </Tooltip>
    );
    const tooltipTrigger = container.querySelector('button');

    expect(container.querySelector('div[role="tooltip"]')).toBeInTheDocument();

    fireEvent.mouseEnter(tooltipTrigger);

    expect(container.querySelector('div[role="tooltip"]')).toBeInTheDocument();

    fireEvent.mouseLeave(tooltipTrigger);

    expect(container.querySelector('div[role="tooltip"]')).toBeInTheDocument();
  });

  it('should show the tooltip on mouseenter and hide it on mouseleave', () => {
    const { container, getByText } = render(
      <Tooltip content={CONTENT_TEXT}>{TRIGGER_ELEMENT}</Tooltip>
    );
    const tooltipTrigger = container.querySelector('button');

    expect(
      container.querySelector('div[role="tooltip"]')
    ).not.toBeInTheDocument();

    fireEvent.mouseEnter(tooltipTrigger);

    expect(container.querySelector('div[role="tooltip"]')).toBeInTheDocument();

    fireEvent.mouseLeave(tooltipTrigger);

    expect(
      container.querySelector('div[role="tooltip"]')
    ).not.toBeInTheDocument();
  });

  it('should hide the tooltip when the escape key is pressed', () => {
    const { container, getByText } = render(
      <Tooltip content={CONTENT_TEXT}>{TRIGGER_ELEMENT}</Tooltip>
    );
    const tooltipTrigger = container.querySelector('button');

    fireEvent.focus(tooltipTrigger);

    const tooltip = container.querySelector('div[role="tooltip"]');

    expect(tooltip).toBeInTheDocument();

    fireEvent.keyDown(tooltipTrigger, {
      key: 'ArrowDown',
      code: 40,
    });

    expect(tooltip).toBeInTheDocument();

    fireEvent.keyDown(tooltipTrigger, {
      key: 'Escape',
      keyCode: 27,
    });

    expect(tooltip).not.toBeInTheDocument();
  });

  it('should render the tooltip component with the correct styles for the inverse prop, position top', async () => {
    const { container, getByText } = render(
      <Tooltip content={CONTENT_TEXT} isInverse>
        {TRIGGER_ELEMENT}
      </Tooltip>
    );
    const tooltipTrigger = container.querySelector('button');

    await act(async () => {
      tooltipTrigger.focus();
    });

    const tooltip = container.querySelector('div[role="tooltip"]');

    expect(tooltip).toHaveStyleRule('background', magma.colors.neutral100);
    expect(tooltip).toHaveStyleRule('color', magma.colors.neutral700);
  });

  it('should render the tooltip component with the correct styles for the inverse prop, positioned bottom', async () => {
    const { container, getByText } = render(
      <Tooltip content={CONTENT_TEXT} position="bottom" isInverse>
        {TRIGGER_ELEMENT}
      </Tooltip>
    );
    const tooltipTrigger = container.querySelector('button');

    await act(async () => {
      tooltipTrigger.focus();
    });

    const tooltip = container.querySelector('div[role="tooltip"]');

    expect(tooltip).toHaveStyleRule('background', magma.colors.neutral100);
    expect(tooltip).toHaveStyleRule('color', magma.colors.neutral700);
  });

  it('should render the tooltip component with the correct styles for the inverse prop, positioned left', async () => {
    const { container, getByText } = render(
      <Tooltip content={CONTENT_TEXT} position="left" isInverse>
        {TRIGGER_ELEMENT}
      </Tooltip>
    );
    const tooltipTrigger = container.querySelector('button');

    await act(async () => {
      tooltipTrigger.focus();
    });

    const tooltip = container.querySelector('div[role="tooltip"]');

    expect(tooltip).toHaveStyleRule('background', magma.colors.neutral100);
    expect(tooltip).toHaveStyleRule('color', magma.colors.neutral700);
  });

  it('should render the tooltip component with the correct styles for the inverse prop, positioned right', async () => {
    const { container, getByText } = render(
      <Tooltip content={CONTENT_TEXT} position="right" isInverse>
        {TRIGGER_ELEMENT}
      </Tooltip>
    );
    const tooltipTrigger = container.querySelector('button');

    await act(async () => {
      tooltipTrigger.focus();
    });

    const tooltip = container.querySelector('div[role="tooltip"]');

    expect(tooltip).toHaveStyleRule('background', magma.colors.neutral100);
    expect(tooltip).toHaveStyleRule('color', magma.colors.neutral700);
  });

  it('should throw an error if the tooltip children is more than one element', () => {
    const originalError = console.error;
    console.error = (...args) => {
      if (
        /Tooltip children can only be one element/.test(args[0]) ||
        /The above error occurred/.test(args[0])
      ) {
        return;
      }
      originalError.call(console, ...args);
    };

    const renderComponent = () =>
      render(
        <Tooltip content="Tooltip content">
          <div>1 child</div>
          <div>2 child</div>
        </Tooltip>
      );

    expect(renderComponent).toThrowError(
      'Tooltip children can only be one element.'
    );

    console.error = originalError;
  });

  it('Does not violate accessibility standards', () => {
    const { container } = render(
      <Tooltip content={CONTENT_TEXT}>{TRIGGER_ELEMENT}</Tooltip>
    );

    return axe(container.innerHTML).then(result => {
      return expect(result).toHaveNoViolations();
    });
  });
});
