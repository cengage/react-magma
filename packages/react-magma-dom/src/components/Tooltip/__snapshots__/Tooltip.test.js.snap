// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip should render the tooltip component, positioned top by default 1`] = `
.emotion-1 {
  background: #707070;
  border-radius: 8px;
  color: #FFFFFF;
  font-size: 12px;
  -webkit-letter-spacing: .32px;
  -moz-letter-spacing: .32px;
  -ms-letter-spacing: .32px;
  letter-spacing: .32px;
  line-height: 16px;
  font-weight: 600;
  max-width: 300px;
  min-height: 2.5em;
  padding: 8px 12px;
  z-index: 999;
}

.emotion-1[data-popper-placement='top'] {
  margin-bottom: 14px;
}

.emotion-1[data-popper-placement='top'] > span:last-child {
  bottom: 10px;
}

.emotion-1[data-popper-placement='bottom'] {
  margin-top: 14px;
}

.emotion-1[data-popper-placement='bottom'] > span:last-child {
  top: 10px;
}

.emotion-1[data-popper-placement='left'] {
  margin-right: 14px;
}

.emotion-1[data-popper-placement='left'] > span:last-child {
  right: 10px;
}

.emotion-1[data-popper-placement='right'] {
  margin-left: 14px;
}

.emotion-1[data-popper-placement='right'] > span:last-child {
  left: 10px;
}

.emotion-0.emotion-0,
.emotion-0.emotion-0:before {
  display: block;
  height: 8px;
  position: absolute;
  width: 8px;
  z-index: -1;
}

.emotion-0.emotion-0::before {
  content: '';
  -webkit-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
  background: #707070;
}

.emotion-1 {
  background: #707070;
  border-radius: 8px;
  color: #FFFFFF;
  font-size: 12px;
  -webkit-letter-spacing: .32px;
  -moz-letter-spacing: .32px;
  -ms-letter-spacing: .32px;
  letter-spacing: .32px;
  line-height: 16px;
  font-weight: 600;
  max-width: 300px;
  min-height: 2.5em;
  padding: 8px 12px;
  z-index: 999;
}

.emotion-1[data-popper-placement='top'] {
  margin-bottom: 14px;
}

.emotion-1[data-popper-placement='top'] > span:last-child {
  bottom: 10px;
}

.emotion-1[data-popper-placement='bottom'] {
  margin-top: 14px;
}

.emotion-1[data-popper-placement='bottom'] > span:last-child {
  top: 10px;
}

.emotion-1[data-popper-placement='left'] {
  margin-right: 14px;
}

.emotion-1[data-popper-placement='left'] > span:last-child {
  right: 10px;
}

.emotion-1[data-popper-placement='right'] {
  margin-left: 14px;
}

.emotion-1[data-popper-placement='right'] > span:last-child {
  left: 10px;
}

.emotion-0.emotion-0,
.emotion-0.emotion-0:before {
  display: block;
  height: 8px;
  position: absolute;
  width: 8px;
  z-index: -1;
}

.emotion-0.emotion-0::before {
  content: '';
  -webkit-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
  background: #707070;
}

<div
  class="emotion-1"
  data-popper-escaped="true"
  data-popper-placement="top"
  data-popper-reference-hidden="true"
  id="tooltipID"
  role="tooltip"
>
  Test Content
  <span
    class="emotion-0"
    style="position: absolute; left: 0px; transform: translate(0px, 0px);"
  />
</div>
`;
